pose_evaluation/metrics/embedding_distance_metric.py

```diff
@@ -0,0 +1,224 @@
+            dtype (torch.dtype): The data type for the tensor.
+
+        Returns:
+            Tensor: Batch tensor representation of the data on the specified device.
+        """
+        if dtype is None:
+            dtype = self.dtype
+
+        # better performance this way, see https://github.com/pytorch/pytorch/issues/13918
+        if isinstance(data, list) and all(isinstance(x, np.ndarray) for x in data):
+            data = np.asanyarray(data)
+
+        if isinstance(data, list) and all(isinstance(x, torch.Tensor) for x in data):
+            # prevents ValueError: only one element tensors can be converted to Python scalars
+            # https://stackoverflow.com/questions/55050717/converting-list-of-tensors-to-tensors-pytorch
+            data = torch.stack(data)
+
+        return st_util._convert_to_batch_tensor(data).to(device=self.device, dtype=dtype)
+
+    def score(
+        self,
+        hypothesis: TensorConvertableType,
+        reference: TensorConvertableType,
+    ) -> Number:
+        """
+        Compute the distance between two embeddings.
+
+        Args:
+            hypothesis (TensorConvertableType): A single embedding vector.
+            reference (TensorConvertableType): Another single embedding vector.
+
+        Returns:
+            Number: The calculated distance.
+
+        """
+
+        return self.score_all(hypothesis, reference).item()
+
+    def score_all(
+        self,
+        hypotheses: Union[List[TensorConvertableType], Tensor],
+        references: Union[List[TensorConvertableType], Tensor],
+        progress_bar: bool = True,
+    ) -> Tensor:
+        """
+        Compute the distance between all hypotheses and all references.
+
+        Expects 2D inputs. If not already Tensors, will attempt to convert them.
+
+        Args:
+            hypotheses (Union[List[TensorConvertableType], Tensor]):
+                List of hypothesis embeddings or a single tensor.
+            references (Union[List[TensorConvertableType], Tensor]):
+                List of reference embeddings or a single tensor.
+            progress_bar (bool): Whether to display a progress bar. (not implemented yet)
+
+        Returns:
+            Tensor: Distance matrix. Row `i` is the distances of `hypotheses[i]` to all rows of `references`.
+                Shape is be NxM, where N is the number of hypotheses, and M is the number of references
+
+        Raises:
+            TypeError: If either hypotheses or references cannot be converted to a batch tensor
+            ValueError: If the specified metric is unsupported.
+        """
+        try:
+            hypotheses = self._to_batch_tensor_on_device(hypotheses)
+            references = self._to_batch_tensor_on_device(references)
+        except RuntimeError as e:
+            raise TypeError(f"Inputs must support conversion to device tensors: {e}") from e
+
+        if self.kind not in self._metric_dispatch:
```

Comment:
this should happen in the constructor time, not when running the metric

Reply:
true